### ISSUE TITLE
"Got unexpected message on stream x"

### ISSUE THREAD

**Issue by LanderN** [paullouisageneau/libdatachannel#652]
Hello! I'm having another issue related to my previous issue #588. 

Some context: we have a server application that uses libdatachannel datachannels for communicating with a web application.
In the web application, we have multiple tabs, which each correspond to a datachannel. When switching to one of these tabs, a datachannel is opened, when switching to another tab, a new datachannel is opened and the previous datachannel is closed. Additionally, there is one "global" datachannel that stays open for all tabs (is never closed).

An issue occurs when switching between tabs: the datachannel doesn't open and the server logs are full of:
```
2022-06-10 16:00:31.940 DEBUG [238838] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 5
2022-06-10 16:00:31.940 WARN  [238835] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.940 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.940 WARN  [238838] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.940 DEBUG [238837] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.940 WARN  [238838] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.940 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238836] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238837] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238838] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238837] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238836] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238835] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238836] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238835] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.942 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
```
(notice the timestamps: there are hundreds of these messages in a single second)

I was able to reproduce the issue with the following program:
```cpp
#include <rtc/rtc.hpp>
#include <zmq.hpp>
#include <thread>

void server()
{
    rtc::PeerConnection pc(rtc::Configuration {});
    pc.onStateChange([&](rtc::PeerConnection::State state) {
        std::cout << "state change: " << state << std::endl;
    });
    pc.onDataChannel([&](auto dc) {
        std::cout << "data channel opened: " << dc->label() << std::endl;
    });

    // RECV OFFER
    zmq::context_t ctx;
    zmq::socket_t sock(ctx, zmq::socket_type::rep);
    sock.bind("tcp://*:1337");
    zmq::message_t msg;
    sock.recv(msg, zmq::recv_flags::none);
    std::string offer(static_cast<char*>(msg.data()), msg.size());

    pc.onLocalDescription([&](const std::string& sdp) {
        // SEND ANSWER
        zmq::message_t msg(sdp.size());
        std::memcpy(msg.data(), sdp.data(), sdp.size());
        sock.send(msg, zmq::send_flags::none);
    });
    pc.setRemoteDescription(offer);

    while(1) {}
}

void client()
{
    rtc::PeerConnection pc(rtc::Configuration {});
    pc.onStateChange([&](rtc::PeerConnection::State state) {
        std::cout << "state change: " << state << std::endl;
    });

    auto dc = pc.createDataChannel("dummy", rtc::DataChannelInit{});
    auto offer = pc.localDescription();

    // SEND OFFER
    zmq::context_t ctx;
    zmq::socket_t sock(ctx, zmq::socket_type::req);
    sock.connect("tcp://localhost:1337");
    sock.send(zmq::buffer(std::string(*offer)), zmq::send_flags::none);
    zmq::message_t msg;
    // RECV ANSWER
    sock.recv(msg, zmq::recv_flags::none);
    auto answer = std::string(static_cast<char*>(msg.data()), msg.size());

    pc.setRemoteDescription(answer);

    bool flip_flop = true;
    std::shared_ptr<rtc::DataChannel> new_dc_ptr;

    // Switch between "foo" and "bar" datachannels
    while(1) {
        auto name = flip_flop ? "foo" : "bar";
        flip_flop = !flip_flop;

        new_dc_ptr = pc.createDataChannel(name, rtc::DataChannelInit{});
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }
}


int main(int argc, char *argv[])
{
    rtc::InitLogger(rtc::LogLevel::Debug);

    // start server (in another thread)
    std::thread server_thread(server);

    // start client
    client();

    return 0;
}
```
(this code depends on ZeroMQ (with cppzmq c++ wrapper), please let me know if it would be easier for you if I provided an example without this dependency)

The full log looks like this:
```
2022-06-10 16:00:30.933 DEBUG [238831] [rtc::impl::Init::doInit@120] Global initialization
2022-06-10 16:00:30.934 DEBUG [238835] [rtc::impl::Certificate::Generate@223] Generating certificate (OpenSSL)
2022-06-10 16:00:30.935 DEBUG [238831] [rtc::impl::IceTransport::IceTransport@58] Initializing ICE transport (libjuice)
2022-06-10 16:00:30.935 DEBUG [238831] [rtc::impl::PeerConnection::processLocalDescription@943] Adding application to local description, mid="0"
2022-06-10 16:00:30.935 INFO  [238831] [rtc::impl::PeerConnection::changeSignalingState@1177] Changed signaling state to have-local-offer
2022-06-10 16:00:30.936 INFO  [238831] [rtc::impl::PeerConnection::changeGatheringState@1164] Changed gathering state to in-progress
2022-06-10 16:00:30.936 INFO  [238831] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to gathering
2022-06-10 16:00:30.937 INFO  [238831] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to connecting
2022-06-10 16:00:30.937 INFO  [238831] [rtc::impl::PeerConnection::changeState@1146] Changed state to connecting
2022-06-10 16:00:30.937 INFO  [238831] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:2269: Candidate gathering done
2022-06-10 16:00:30.937 INFO  [238831] [rtc::impl::PeerConnection::changeGatheringState@1164] Changed gathering state to complete
state change: connecting
2022-06-10 16:00:30.937 DEBUG [238838] [rtc::impl::Certificate::Generate@223] Generating certificate (OpenSSL)
2022-06-10 16:00:30.938 DEBUG [238832] [rtc::impl::IceTransport::IceTransport@58] Initializing ICE transport (libjuice)
2022-06-10 16:00:30.938 INFO  [238832] [rtc::impl::PeerConnection::changeSignalingState@1177] Changed signaling state to have-remote-offer
2022-06-10 16:00:30.938 DEBUG [238832] [rtc::impl::PeerConnection::processLocalDescription@861] Reciprocating application in local description, mid="0"
2022-06-10 16:00:30.938 INFO  [238832] [rtc::impl::PeerConnection::changeSignalingState@1177] Changed signaling state to stable
2022-06-10 16:00:30.938 INFO  [238832] [rtc::impl::PeerConnection::changeGatheringState@1164] Changed gathering state to in-progress
2022-06-10 16:00:30.938 INFO  [238832] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to gathering
2022-06-10 16:00:30.939 INFO  [238832] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to connecting
2022-06-10 16:00:30.939 INFO  [238832] [rtc::impl::PeerConnection::changeState@1146] Changed state to connecting
state change: connecting
2022-06-10 16:00:30.939 INFO  [238832] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:2269: Candidate gathering done
2022-06-10 16:00:30.939 INFO  [238831] [rtc::impl::PeerConnection::changeSignalingState@1177] Changed signaling state to stable
2022-06-10 16:00:30.939 INFO  [238832] [rtc::impl::PeerConnection::changeGatheringState@1164] Changed gathering state to complete
2022-06-10 16:00:30.939 DEBUG [238831] [rtc::PeerConnection::setLocalDescription@109] No negotiation needed
2022-06-10 16:00:30.939 INFO  [238843] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to connected
2022-06-10 16:00:30.939 DEBUG [238843] [rtc::impl::DtlsTransport::DtlsTransport@383] Initializing DTLS transport (OpenSSL)
2022-06-10 16:00:30.940 DEBUG [238843] [rtc::impl::DtlsTransport::start@472] Starting DTLS recv thread
2022-06-10 16:00:30.940 INFO  [238843] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to connected
2022-06-10 16:00:30.940 DEBUG [238843] [rtc::impl::DtlsTransport::DtlsTransport@383] Initializing DTLS transport (OpenSSL)
2022-06-10 16:00:30.940 DEBUG [238843] [rtc::impl::DtlsTransport::start@472] Starting DTLS recv thread
2022-06-10 16:00:30.940 INFO  [238843] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to completed
2022-06-10 16:00:30.940 INFO  [238843] [rtc::impl::IceTransport::LogCallback@354] juice: agent.c:1002: Changing state to completed
2022-06-10 16:00:30.941 INFO  [238845] [rtc::impl::DtlsTransport::runRecvLoop@565] DTLS handshake finished
2022-06-10 16:00:30.941 INFO  [238844] [rtc::impl::DtlsTransport::runRecvLoop@565] DTLS handshake finished
2022-06-10 16:00:30.941 DEBUG [238845] [rtc::impl::SctpTransport::SctpTransport@188] Initializing SCTP transport
2022-06-10 16:00:30.941 DEBUG [238844] [rtc::impl::SctpTransport::SctpTransport@188] Initializing SCTP transport
2022-06-10 16:00:30.942 DEBUG [238844] [rtc::impl::SctpTransport::connect@370] SCTP connecting (local port=5000, remote port=5000)
2022-06-10 16:00:30.942 DEBUG [238845] [rtc::impl::SctpTransport::connect@370] SCTP connecting (local port=5000, remote port=5000)
2022-06-10 16:00:30.942 DEBUG [238837] [rtc::impl::SctpTransport::processNotification@824] SCTP negotiated streams: incoming=1024, outgoing=1024
2022-06-10 16:00:30.942 DEBUG [238836] [rtc::impl::SctpTransport::processNotification@824] SCTP negotiated streams: incoming=1024, outgoing=1024
2022-06-10 16:00:30.943 INFO  [238837] [rtc::impl::SctpTransport::processNotification@829] SCTP connected
2022-06-10 16:00:30.943 INFO  [238836] [rtc::impl::SctpTransport::processNotification@829] SCTP connected
2022-06-10 16:00:30.943 INFO  [238837] [rtc::impl::PeerConnection::changeState@1146] Changed state to connected
2022-06-10 16:00:30.943 INFO  [238836] [rtc::impl::PeerConnection::changeState@1146] Changed state to connected
state change: connected
state change: connected
data channel opened: foo
2022-06-10 16:00:30.943 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
data channel opened: dummy
2022-06-10 16:00:30.944 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 1
2022-06-10 16:00:31.939 DEBUG [238831] [rtc::PeerConnection::setLocalDescription@109] No negotiation needed
2022-06-10 16:00:31.939 DEBUG [238837] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
data channel opened: bar
2022-06-10 16:00:31.940 DEBUG [238838] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 5
2022-06-10 16:00:31.940 WARN  [238835] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.940 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.940 WARN  [238838] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.940 DEBUG [238837] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.940 WARN  [238838] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.940 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238836] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238837] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238838] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238837] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238836] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238835] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.941 DEBUG [238836] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
2022-06-10 16:00:31.941 WARN  [238835] [rtc::impl::PeerConnection::forwardMessage@459] Got unexpected message on stream 3
2022-06-10 16:00:31.942 DEBUG [238835] [rtc::impl::SctpTransport::sendReset@682] SCTP resetting stream 3
[... repeats indefinitely ...]
```

Obviously, this situation looks contrived on its own, but at least it demonstrates the problem :)

If it helps, I'm using this [CMakeLists.txt](https://github.com/paullouisageneau/libdatachannel/files/8879663/CMakeLists.txt) for compiling the example.

Please let me know if you need any additional information. Thanks!

**Comment by paullouisageneau** [comment]
Thank you for the very detailed report and the code to reproduce!

It looks like this particular scenario triggers a ping-pong of close messages which then prevents from opening the new data channel.

**Comment by LanderN** [comment]
That was fast! Thanks for looking into this, I will be able to test this out and will report back on Monday.

**Comment by LanderN** [comment]
I pulled in the code from the pull request and unfortunately, the problem persists: datachannels don't seem to open and the log is full of `Got unexpected message on stream` messages.

I've reworked the example code to get rid of the ZeroMQ dependency:
```cpp
#include <rtc/rtc.hpp>
#include <functional>
#include <thread>

void run_server(rtc::PeerConnection& pc,
                std::string offer,
                std::function<void(std::string)> answer_cb)
{
    pc.onStateChange([&](rtc::PeerConnection::State state) {
        std::cout << "state change: " << state << std::endl;
    });
    pc.onDataChannel([&](auto dc) {
        std::cout << "data channel opened: " << dc->label() << std::endl;
    });

    // RECV OFFER
    pc.onLocalDescription([&](const std::string& sdp) {
        // SEND ANSWER
        answer_cb(sdp);
    });
    pc.setRemoteDescription(offer);

    while(1) {}
}

std::string init_client(rtc::PeerConnection& pc)
{
    pc.onStateChange([&](rtc::PeerConnection::State state) {
        std::cout << "state change: " << state << std::endl;
    });

    auto dc = pc.createDataChannel("dummy", rtc::DataChannelInit{});
    auto offer = pc.localDescription();

    return *offer;
}

void run_client(rtc::PeerConnection& pc)
{
    while (!pc.remoteDescription()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }

    bool flip_flop = true;
    std::shared_ptr<rtc::DataChannel> new_dc_ptr;

    // Switch between "foo" and "bar" datachannels
    while(1) {
        auto name = flip_flop ? "foo" : "bar";
        flip_flop = !flip_flop;

        new_dc_ptr = pc.createDataChannel(name, rtc::DataChannelInit{});
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }
}


int main(int argc, char *argv[])
{
    rtc::InitLogger(rtc::LogLevel::Debug);

    // start client
    rtc::PeerConnection client_pc(rtc::Configuration {});
    auto offer = init_client(client_pc);

    // start server (in another thread)
    rtc::PeerConnection server_pc(rtc::Configuration {});
    std::thread server_thread(run_server, std::ref(server_pc), offer, [&](const std::string& answer){
        client_pc.setRemoteDescription(answer);
    });

    run_client(client_pc);

    return 0;
}
```

**Comment by paullouisageneau** [comment]
@LanderN Sorry, it is properly fixed now.

Thank you for the reworked example, I'll add an adapted version as a test if you don't mind.

**Comment by LanderN** [comment]
Works perfectly now, thanks! A test based on this use case sounds great as well :+1:

**Comment by paullouisageneau** [comment]
Fix released in 0.17.7
